vpr/src/route/rr_graph_fpga_interchange.cpp

```diff
@@ -0,0 +1,1706 @@
+                        }
+                    }
+                }
+                for (auto i : NODESIDES)
+                    single_node &= !vertex->links[i];
+                if (!vertex->has_pins && single_node)
+                    cnt++;
+            }
+        }
+        return cnt;
+    }
+
+    void process_set(std::unordered_set<intermediate_node*>& set,
+                     std::unordered_set<location, hash_tuple::hash<location>>& nodes_used,
+                     std::map<location, intermediate_node*>& existing_nodes,
+                     std::map<location, std::tuple<location, e_rr_type, int>>& local_redirect,
+                     float R,
+                     float C,
+                     location offset,
+                     node_sides side,
+                     e_rr_type chan_type) {
+        int len;
+        int idx;
+        intermediate_node *start, *end;
+        for (auto const i : set) {
+            len = 0;
+            start = i;
+            end = i;
+            auto key = std::make_tuple(add_vec(start->loc, offset), chan_type);
+            idx = virtual_chan_loc_map_[key].size();
+            do {
+                nodes_used.insert(end->loc);
+                len++;
+                local_redirect.emplace(end->loc, std::make_tuple(add_vec(start->loc, offset), chan_type, idx));
+                if (!end->links[side])
+                    break;
+                intermediate_node* neighbour = existing_nodes[add_vec(end->loc, offsets[side])];
+                if (set.find(neighbour) != set.end())
+                    break;
+                end = neighbour;
+            } while (true);
+            virtual_chan_loc_map_[key].emplace_back(start->segment_id, R * len, C * len, add_vec(end->loc, offset));
+        }
+    }
+
+    void add_short(location n1,
+                   location n2,
+                   std::map<location, std::tuple<location, e_rr_type, int>>& r1,
+                   std::map<location, std::tuple<location, e_rr_type, int>>& r2) {
+        auto red1 = r1[n1];
+        auto red2 = r2[n2];
+        virtual_shorts_.emplace_back(std::get<0>(red1), std::get<1>(red1), std::get<2>(red1),
+                                     std::get<0>(red2), std::get<1>(red2), std::get<2>(red2));
+    }
+
+    void connect_base_on_redirects(std::unordered_set<intermediate_node*>& set,
+                                   node_sides side,
+                                   std::map<location, std::tuple<location, e_rr_type, int>>& src,
+                                   std::map<location, std::tuple<location, e_rr_type, int>>& dist) {
+        for (auto const node : set) {
+            if (!node->links[side])
+                continue;
+            location other_node = add_vec(node->loc, offsets[side]);
+            if (dist.find(other_node) == dist.end())
+                continue;
+            add_short(node->loc, other_node, src, dist);
+        }
+    }
+
+    void graph_reduction_stage2(int node_id,
```

Comment:
Is there a first reduction stage? Otherwise I think this could be called only `graph_reduction_stage`

Reply:
There is in fact stage one, but it has name `educe_graph_and_count_nodes_left`